[SYCL] Builtin support + missing changes

elizabethandrews opened the pull request:
Commit 1 - Fixes a bug to generate kernel caller function correctly when kernel entry point attribute is not a template
Commits 2 and 3 - Builtin support
Commit - 4 - Fixes discrepancies between Tom and Elizabeth forks

elizabethandrews committed:
[SYCL] Do not emit SYCL kernel entry point function eagerly

In SYCL device compilation, SYCL kernel entry point functions
trigger the generation of SYCL kernel caller functions. This
is done when handling deferred emissions.

elizabethandrews committed:
[SYCL] Add support for builtin-sycl-kernel-param-count.cpp

elizabethandrews committed:
[SYCL] Add support for __builtin_sycl_kernel_name

The builtin takes the kernel name type as it's argument
and returns the mangled name for the kernel caller
function.

elizabethandrews committed:
[SYCL] Add missing changes in Tom's fork

tahonermann commented on clang/lib/CodeGen/CodeGenSYCL.cpp:
```diff
@@ -45,12 +45,14 @@ void CodeGenModule::EmitSYCLKernelCaller(const FunctionDecl *KernelEntryPointFn,
-  assert(SKEPAttr && "Missing sycl_kernel_entry_point attribute");
-  CanQualType KernelNameType = Ctx.getCanonicalType(SKEPAttr->getKernelName());
-  const SYCLKernelInfo &SKI = Ctx.SYCLKernels.at(KernelNameType);
+  const auto *KernelEntryPointAttr =
+      KernelEntryPointFn->getAttr<SYCLKernelEntryPointAttr>();
```
Any reason not to preserve the assert that was there?
```suggestion
  const auto *KernelEntryPointAttr =
      KernelEntryPointFn->getAttr<SYCLKernelEntryPointAttr>();
  assert(KernelEntryPointAttr && "Missing sycl_kernel_entry_point attribute");
```

elizabethandrews replied:
I didn't think it was necessary because we enter this function only if the attribute is present.

tahonermann replied:
That is true, but the motivation for use of assert is to ensure that such relationships are sustained during code refactoring. In this context, the assert is used as a function precondition (C++ contracts will provide a better way to declare such requirements in the future; I'm on the hook to deliver a presentation on C++ contracts sometime this year). Granted, in this case, if a caller neglects to ensure that the function actually has the attribute, we'll likely get a SIGSEGV, but an assertion failure generally produces a better understanding of what went wrong.

tahonermann commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -2450,6 +2450,25 @@ static bool BuiltinCountZeroBitsGeneric(Sema &S, CallExpr *TheCall) {
+  IdentifierTable &IdentTable = S.Context.Idents;
+  auto Name = DeclarationName(&(IdentTable.get("type")));
+  DeclContext::lookup_result Lookup = RD->lookup(Name);
+  if (Lookup.empty() || !Lookup.isSingleResult() || !isa<TypeAliasDecl>(Lookup.front()))
```
```suggestion
  if (Lookup.empty() || !Lookup.isSingleResult() || !isa<TypedefNameDecl>(Lookup.front()))
```

tahonermann commented on clang/test/SemaSYCL/builtin-sycl-kernel-name.cpp:
```diff
@@ -0,0 +1,44 @@
+  const char* test5 = __builtin_sycl_kernel_name("str"); // expected-error {{invalid argument; argument must be a class or struct type with a member type alias named 'type'}}
+  const char* test6 = __builtin_sycl_kernel_name(kernel_id_2(), kernel_id_2()); // expected-error {{builtin requires exactly 1 argument}}
+}
+
```
Additional test suggestions:

Test invocation in constant evaluation:
```
struct constexpr_kernel_name1;
void test_constexpr() {
  kernel_single_task<constexpr_kernel_name1>([]{});
  constexpr const char* test1 = __builtin_sycl_kernel_name(kernel_id_1<constexpr_kernel_name1>()); // Valid
}
```

Test that a non-constant argument produces an error.
```
struct nonconst_kernel_name1;
struct nonconst_kernel_name2;
void test() {
  kernel_single_task<nonconst_kernel_name1>([]{});
  kernel_single_task<nonconst_kernel_name2>([]{});
  constexpr kernel_id_1<nonconst_kernel_name1> kn1;
  (void)__builtin_sycl_kernel_name(kn1); // Valid
  kernel_id_1<nonconst_kernel_name2> kn2;
  (void)__builtin_sycl_kernel_name(kn2); // expected-error {{FIXME}}
}
```

tahonermann replied:
Additional cases to exercise:
- Calling the builtin in a constexpr context before a call to `kernel_single_task()`; ensure an error is produced.
- Calling the builtin before a call to `kernel_single_task()`; this should work by accessing data in a symbol declaration for which a symbol definition is emitted sometime after the call to `kernel_single_task()` is observed (perhaps at the end of the translation unit contingent on use of the symbol).
- Calling the builtin with no corresponding call to `kernel_single_task()`; ensure an error is produced at the end of the translation unit. This case is similar to the previous case; the error occurs because the symbol definition is never emitted.

tahonermann reviewed (changes requested):
Thanks @elizabethandrews! The changes look good. Most of my suggestions are for changes to comments or for additional tests. I did not a couple of technical concerns as well, but overall this looks right.

elizabethandrews commented on clang/lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -3197,21 +3213,6 @@ void CodeGenModule::EmitDeferred() {
     if (LangOpts.OpenMP && OpenMPRuntime && OpenMPRuntime->emitTargetGlobal(D))
       continue;
 
-    // If the Decl corresponds to a SYCL kernel entry point function, generate
-    // and emit the corresponding SYCL kernel caller function, i.e the
-    // offload kernel. Otherwise, emit the definition and move on to the next
-    // one.
-    const FunctionDecl *FD = nullptr;
-    if (LangOpts.SYCLIsDevice &&
-        (FD = D.getDecl()->getAsFunction()) != nullptr &&
-        FD->hasAttr<SYCLKernelEntryPointAttr>() &&
-        FD->isDefined()) {
-      // Generate and emit the offload kernel
-      EmitSYCLKernelCaller(FD, getContext());
-    } else {
-      EmitGlobalDefinition(D, GV);
-    }
-
```
Oh I think I accidently removed that. It is there in my fork. I'll put it back!

elizabethandrews commented on clang/lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -3168,6 +3168,22 @@ void CodeGenModule::EmitDeferred() {
   CurDeclsToEmit.swap(DeferredDeclsToEmit);
 
   for (GlobalDecl &D : CurDeclsToEmit) {
+    // If the Decl corresponds to a SYCL kernel entry point function, generate
+    // and emit the corresponding the SYCL kernel caller function i.e the
+    // offload kernel.
+    if (const auto *FD = D.getDecl()->getAsFunction()) {
+      if (LangOpts.SYCLIsDevice && FD->hasAttr<SYCLKernelEntryPointAttr>() &&
+          FD->isDefined()) {
+        // Generate and emit the offload kernel
+        EmitSYCLKernelCaller(FD, getContext());
+        // The offload kernel invokes the operator method of the SYCL kernel
+        // object i.e. the SYCL kernel function is invoked. Emit this function.
+        EmitDeferred();
+        // Do not emit the SYCL kernel entry point function.
+        continue;
+      }
+    }
+
```
The issue is GetAddrOfGlobal below will result in an emit of the entry-point function. We need this block in the beginning to avoid emitting anything related the the entry-point function.

tahonermann replied:
Ah, I didn't realize that. That makes sense. Perhaps we should add a comment to that effect. And update tests to validate that the entry point function isn't emitted on the device side.

elizabethandrews commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -13820,9 +13826,38 @@ static SYCLKernelInfo BuildSYCLKernelInfo(ASTContext &Context,
+}
+
+static int GetSYCLKernelCallerParamCount(const FunctionDecl *FD) {
+  // The parameters of the compiler generated SYCL kernel caller function
+  // are generated using the parameters of the SYCL kernel entry point
+  // function. The first parameter of the SYCL kernel entry point function
+  // (i.e. FD), is the SYCL kernel object. If the SYCL kernel object does
+  // not contain SYCL special types, it can be passed as-is to the device.
+  // In this case, parameters of the SYCL kernel caller function are the
+  // same as that of the the SYCL kernel entry point function. However, if
+  // the SYCL kernel object contains a SYCL special type, it is decomposed
+  // to it's data members and the parameters of the kernel caller function
+  // are generated using these decomposed fields, instead of the kernel
+  // object.
+
+  // FIXME: SYCL special types are not currently supported and therefore we
+  // assume there is no decomposition and return the parameter count of SYCL
+  // kernel entry point function.
```
Will we be upstreaming builtins in initial PR? I thought this was just for internal testing. If we are upstreaming all this initially I assume I don't even need this function and can just do `FD->param_size()` in `BuildSYCLKernelInfo`

tahonermann replied:
Good question. My memory is fuzzy on that. Let's talk about that during the FE sync meeting today.

elizabethandrews committed:
[SYCL] Fix incorrect merge with Tom's fork

elizabethandrews commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -12199,6 +12199,9 @@ def err_sycl_kernel_name_type : Error<
   "'sycl_kernel_entry_point' kernel name argument conflicts with a previous"
   " declaration">;
+def err_builtin_invalid_argument_count : Error<"builtin requires exactly 1 argument">;
+def err_builtin_invalid_argument : Error<"invalid argument; argument must be a class or struct type"
+                                         " with a member type alias named 'type'">;
```
I took inspiration from some existing diagnostics with "argument must be a xyz" wording. I personally prefer that to "expected a...". I have no strong feelings though and so I have made this modification. 

I did not specify the type via %0 in the diagnostic because we have the ^ pointing to the problematic type when we emit the diagnostic. Way back, someone told me to not include unnecessary information in diagnostics to keep it short, if we are already pointing to the problematic type.

tahonermann replied:
I'm fine with the "argument must be a xyz" wording; that is consistent with other diagnostics too.

My motivation for including the type in the diagnostic is because the carat won't actually point to the type; I expect it will point to the expression that produces an object of the type. Given this:
```
template<typename KernelName>
struct kernel_id_t {
  using type = KernelName;
};
template<typename KernelName>
void f() {
  __builtin_sycl_kernel_name(kernel_id_t<KernelName>());
}
```

I would expect a diagnostic to look something like:
```
invalid argument; argument must be a class or structure with a member typedef or type alias named 'type'
  __builtin_sycl_kernel_name(kernel_id_t<KernelName>());
                             ^^^^^^^^^^^^^^^^^^^^^^^^^
```

Our expectation is that the expression will readily identify the type in expected usage as in the above case, but that isn't guaranteed. For example, the expression could call some function (I don't know of any good reason to do so, but style preferences could lead someone to do so).

elizabethandrews commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -2538,6 +2539,20 @@ static RValue EmitHipStdParUnsupportedBuiltin(CodeGenFunction *CGF,
 }
 
+const SYCLKernelInfo *GetSYCLKernelInfo(ASTContext &Ctx, const CallExpr *E) {
+  // Argument to the builtin is a kernel_id_t type trait which is used
+  // to retrieve the kernel name type.
```
Can you please explain why we need it to be default constructible and copy constructible?

elizabethandrews replied:
Actually can you explain this comment? How is the kernel identified by the argument's type? Won't the type of the argument be the class/struct enclosing type alias named type, not the kernel name type itself?

tahonermann replied:
Calling the builtin requires constructing an object of the type and passing it as an argument to the builtin. However, you are right that we don't need to be overly prescriptive here; from the perspective of the builtin, it doesn't matter how the object is constructed.

Upon re-reading my suggested update, I agree some of it is ambiguous regarding which type it is referring to. How does this sound?
```suggestion
  // The first argument to the builtin is an object that designates the SYCL kernel.
  // The argument is evaluated and its value is discarded; the SYCL kernel is
  // identified based on the argument type. The argument type is required to be
  // a class or structure with a member typedef or type alias named 'type'. The
  // target type of the 'type' member is the SYCL kernel name type.
```

elizabethandrews committed:
[SYCL] Apply review comments - Builtins

elizabethandrews committed:
[SYCL] Apply review comments - kernel generation

elizabethandrews committed:
[SYCL] Implement builtins

Implemented support for __builtin_sycl_kernel_param_kind,
__builtin_sycl_kernel_param_size and __builtin_sycl_kernel_param_offset

Changes cause ast-dump-sycl-kernel-call-stmt.cpp to fail. The failure
occurs when we try to obtain the size of a parameter with an
instantiated type. The failure is because the definition of the type
is not instantiated when attempting to obtain it's size. It is
unclear why this is happening. Maybe KernelInfo is being collected
too early? Further debugging is required.

elizabethandrews committed:
[SYCL] Dummy implementation of __builtin_sycl_kernel_param_access_target

This is a dummy implementation so that the library will compile.
Actual implementation will be done after SYCL special types
are supported.

elizabethandrews committed:
[SYCL] Handling of kernel entry point parameters should be generic

The old design required the SYCL kernel object to be the first
parameter of the kernel entry point function. In the new design,
all parameters of the kernel entry point function are treated
in a similar manner. We do not have a designated spot for the
kernel object.

elizabethandrews committed:
[SYCL] Simplify implementation of __builtin_sycl_kernel_param_count

elizabethandrews committed:
[SYCL] Fix crashing test

Getting the size of parameter requires it to be a complete type.
When the parameter sizes were previously stored in SYCLKernelInfo,
the compiler hit an assert in cases where the definition of the type
was not yet instantiated by the compiler. This PR moves the code
calculating size to the point of invocation of the builtin (instead
of kernel invocation), hoping the type is fully instantiated by the
compiler by then. This needs to be verified by adding new builtin
tests. The existing crashing test now passes because it doesn't
invoke the builtin.

elizabethandrews committed:
[SYCL] Dummy implementation of code location builtins

The following builtins were implemented to point to the
location of the SYCL kernel object. In the new design,
the SYCL kernel object does not having 'special'
designation since all members of the entry point function
will be treated in a similar manner. It is unclear what
these builtins should now return or whether they should
even exist. This PR is a dummy implementation to make
library compile while we investigate.

__builtin_sycl_kernel_file_name(kernel-name)
__builtin_sycl_kernel_function_name(kernel-name)
__builtin_sycl_kernel_line_number(kernel-name)
__builtin_sycl_kernel_column_number(kernel-name)

tahonermann commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -2450,6 +2450,26 @@ static bool BuiltinCountZeroBitsGeneric(Sema &S, CallExpr *TheCall) {
   return false;
 }
 
+// The argument must be a class or struct with a member
+// named type.
+static bool CheckBuiltinSyclKernelName(Sema &S, CallExpr *TheCall) {
+  QualType ArgTy = TheCall->getArg(0)->getType();
+  const auto *RT = ArgTy->getAs<RecordType>();
```
The primary intent of this function appears to be to validate that the type that identifies the kernel meets specific requirements. This is useful and is separable from the fact that the type is obtained by querying the type of the first argument in a call expression; I think it would be clearer to colocate the retrieval of the type with the code that checks for the number of arguments below.
```suggestion
// The SYCL kernel builtin functions identify the kernel to operate on
// by the type of one of the arguments. That type is required to be a
// class or structure type with a member typedef or type alias named
// "type".
static bool CheckBuiltinSyclKernelIDType(Sema &S, QualType T) {
  const auto *RT = T->getAs<RecordType>();
```

tahonermann commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -2450,6 +2450,26 @@ static bool BuiltinCountZeroBitsGeneric(Sema &S, CallExpr *TheCall) {
+
+  RecordDecl *RD = RT->getDecl();
+  IdentifierTable &IdentTable = S.Context.Idents;
+  auto Name = DeclarationName(&(IdentTable.get("type")));
```
I don't know if I should be worried about this, but `IdentifierTable::get()` will create an `IdentifierInfo` object if one doesn't already exist. That makes me nervous. There is a `find()` member function to do a lookup without creating a new object, but it isn't particularly ergonomic to use. The lack of a `getExisting()` or similar function that can return null is a little surprising to me though, so maybe I'm being overly paranoid. Maybe it is worth using `find()` here to avoid unlikely potential problems?

tahonermann commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -3252,6 +3272,55 @@ Sema::CheckBuiltinFunctionCall(FunctionDecl *FDecl, unsigned BuiltinID,
+      return ExprError();
+    }
+
+    if (CheckBuiltinSyclKernelName(*this, TheCall)) {
```
Per prior comment:
```suggestion
    if (CheckBuiltinSyclKernelName(*this, TheCall->getArg(0)->getType())) {
```

tahonermann commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -3252,6 +3272,55 @@ Sema::CheckBuiltinFunctionCall(FunctionDecl *FDecl, unsigned BuiltinID,
+  case Builtin::BI__builtin_sycl_kernel_param_offset:
+  case Builtin::BI__builtin_sycl_kernel_param_size:
+  case Builtin::BI__builtin_sycl_kernel_param_access_target: {
+    // Builtin takes 1 argument
```
```suggestion
    // Builtin takes 2 arguments
```

tahonermann reviewed (changes requested):
I requested some minor changes; nothing fancy.

I still need to re-review the tests, but ran out of time for today. I'll look at those tomorrow.

elizabethandrews commented on clang/include/clang/AST/SYCLKernelInfo.h:
```diff
@@ -43,10 +53,33 @@ class SYCLKernelInfo {
 private:
+  // Kernel caller function parameter descriptor.
+  struct KernelParamDesc {
+    kernel_param_kind_t Kind = kind_last;
```
I do not have a good answer here. I used the same implementation of the enum in syclos.

tahonermann commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -5992,6 +6009,81 @@ RValue CodeGenFunction::EmitBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
+    // FIXME: This is a dummy value. This builtin will be implemented when
+    // support for special sycl types is implemented.
+    return RValue::get(llvm::ConstantInt::get(Int32Ty, 0));
+  }
```
For the above builtins that take a parameter index as an argument, we'll need to support the parameter index argument being a non-constant; see how Sergey implemented `getKernelParamDescs()` in https://github.com/intel/llvm/pull/15070. Code consuming the parameter information will look something like:
```
  int param_count = __builtin_sycl_kernel_param_count(KernelIdentity<KernelNameType>());
  for (int i = 0; i < param_count; ++i) {
    int param_size = __builtin_sycl_kernel_param_size(KernelIdentity<KernelNameType>(), i);
    ...
  }
```

There are a few implications of this.

1. When the parameter index argument can be evaluated at compile-time, IR can be emitted to just provide the parameter info value as the code does now. However, we need to issue a diagnostic if the parameter index is not in range.
2. When the parameter index argument *can't* be evaluated at compile-time, IR will have to be emitted to retrieve the value from a stored array of elements. This will presumably require emitting something like a static local constexpr variable (something with internal linkage) that the IR will index into. We can't issue a diagnostic for an out of range parameter index in this case of course, so the IR should include a comparison and branch to an `unreachable` instruction.

tahonermann commented on clang/test/CodeGenSYCL/builtin-sycl-kernel-param-kind.cpp:
```diff
@@ -0,0 +1,55 @@
+// CHECK: store i32 1, ptr %test1, align 4
+// CHECK: store i32 1, ptr %test2, align 4
+// CHECK: store i32 1, ptr %test3, align 4
+
```
Per the comment about the parameter index argument not necessarily being a constant, we should have a test that looks something like this:
```
class kernel_name_3;
void test2() {
  SYCLKernel Obj;
  kernel_single_task<kernel_name_3>(Obj, 1);
  int kn3_param_count = __builtin_sycl_kernel_param_count(kernel_id_t<kernel_name_3>());
  for (int i = 0; i < kn3_param_count; ++i) {
    int param_kind = __builtin_sycl_kernel_param_kind(kernel_id_t<kernel_name_3>(), i);
  }
}
```

I think it would be reasonable to have a single test that covers all of the builtins that take a parameter index argument.

tahonermann reviewed (commented):
Added a couple of comments regarding the parameter index argument potentially being a run-time supplied value.

tahonermann commented:
Closing without merge; we decided to proceed with a different kernel launch mechanism that avoids the need for these builtin functions.
